### PR TITLE
CXX-527 provide an API for BSON validation

### PR DESCRIPTION
**Pull request by amidvidy**
this is for a review of interface and implementation. Docs and tests will be added to this PR in a forthcoming commit.

### CODE REVIEW

**Comment by amidvidy** [comment]
another question is if its worth PIMPLing the validator.

**Comment by amidvidy** [comment]
added docs.

File: src/bsoncxx/stdx/optional.hpp [hunk 1]
```diff
@@ -29,6 +29,7 @@ namespace stdx {
 
 using ::core::optional;
 using ::core::nullopt;
+using ::core::make_optional;
```
Comment (acmorrow) [inline src/bsoncxx/stdx/optional.hpp]: Neat

File: src/bsoncxx/validate.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+/// @param validator
+///   A validator used to configure what checks are done. If the document if found
+///   to be invalid, it will contain the offset at which the document was found to
+///   be invalid.
+///
+/// @returns
+///   An engaged optional containing a view if the document is valid, or
+///   an unengaged optional if the document is invalid.
+///
+BSONCXX_API stdx::optional<document::view> validate(const std::uint8_t* data,
```
Comment (acmorrow) [inline src/bsoncxx/validate.hpp]: I see why you wanted variant. So you could return either a document view, or an object describing the failed validation.

File: src/bsoncxx/validate.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    ///   If true, UTF-8 validation (with null bytes allowed) is performed.
+    ///
+    void check_utf8_allow_null(bool check_utf8_allow_null);
+
+    ///
+    /// Verifies that document keys are not preceeded with '$'. Such keys
+    /// are reserved for MongoDB internal use.
+    ///
+    /// @param check_dollar_keys
+    ///   If true, UTF-8 validation (with null bytes allowed) is performed.
```
Comment (amidvidy) [inline src/bsoncxx/validate.hpp]: bad copy pasta

File: src/bsoncxx/validate.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    ///
+    void check_dollar_keys(bool check_dollar_keys);
+
+    ///
+    /// Verifies that document keys do not contain any '.' characters. The dot
+    /// character is illegal in MongoDB keys.
+    ///
+    /// @param check_dollar_keys
+    ///   If true, UTF-8 validation (with null bytes allowed) is performed.
```
Comment (amidvidy) [inline src/bsoncxx/validate.hpp]: ditto

File: src/bsoncxx/validate.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+/// precise validation that is performed, and will give the caller access to the
+/// offset at which the document was found to be invalid.
+///
+/// @param data
+///   A buffer containing a BSON document to validate.
+/// @param length
+///   The size of the buffer.
+/// @param validator
+///   A validator used to configure what checks are done. If the document if found
+///   to be invalid, it will contain the offset at which the document was found to
```
Comment (amidvidy) [inline src/bsoncxx/validate.hpp]: less repetitive wording

File: src/bsoncxx/validate.hpp (hunk 2 above)
Comment (amidvidy) [inline src/bsoncxx/validate.hpp]: yeah. expected<T, E> would be better, but the [mnmlstc version](http://mnmlstc.github.io/core/optional.html#expected:T:) differs from the version [proposed for the standard](https://isocpp.org/files/papers/n4015.pdf).

**Comment by amidvidy** [comment]
added tests. @acmorrow care to give this a quick look?

File: src/bsoncxx/test/bson_validate.cpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+// These are basic sanity tests to make sure the API works.
+// Libbson has a much more complete suite of tests that covers various corrupt BSON
+// inputs.
+
+namespace {
+    // polymorphic lambdas would be nice here.
+    template<typename T> 
+    bool is_engaged(const stdx::optional<T>& opt) {
+        return opt != stdx::nullopt;
```
Comment (acmorrow) [inline src/bsoncxx/test/bson_validate.cpp]: Can you just return static_cast<bool>(opt) to explictly invoke the explicit conversion to bool?

File: src/bsoncxx/test/bson_validate.cpp [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    
+    template<typename T>
+    bool is_disengaged(const stdx::optional<T>& opt) {
+        return opt == stdx::nullopt;
+    }
+}
```
Comment (acmorrow) [inline src/bsoncxx/test/bson_validate.cpp]: // namespace

**Comment by acmorrow** [comment]
Something of the design of this seems like it could be improved, to me. In particular, the fact that we need to friend in the validate function to the validator, and that the validator is stateful both seem not quite right. I think part of it is that you don't have the Either<T, U> return type.

I think there is a simpler design available, where the validator doesn't own the offset state, and is constant w.r.t. performing validation.

**Comment by amidvidy** [comment]
we should probably discuss this one IRL tomorrow.

File: src/bsoncxx/validate.hpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+///   A validator used to configure what checks are done. If validation fails, it
+///   will contain the offset at which the document was found to be invalid.
+///
+/// @returns
+///   An engaged optional containing a view if the document is valid, or
+///   an unengaged optional if the document is invalid.
+///
+BSONCXX_API stdx::optional<document::view> validate(const std::uint8_t* data,
+                                                    std::size_t length,
+                                                    validator& validator);
```
Comment (acmorrow) [inline src/bsoncxx/validate.hpp]: Lets make the invalid offset an out parameter:

BSONCXX_API stdx::optional<document::view> validate(
    const std::uint8_t* data,
    std::size_t length,
    const validator& validator = validator(),
    std::size_t** invalid_offset = NULL)

And then we can remove the friendship, and the member state.

File: src/bsoncxx/validate.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+stdx::optional<document::view> validate(const std::uint8_t* data, std::size_t length) {
+    validator vtor{};
+    return validate(data, length, vtor);
+}
+
+stdx::optional<document::view> validate(const std::uint8_t* data, std::size_t length, validator& validator) {
+    ::bson_validate_flags_t flags = BSON_VALIDATE_NONE;
+
+    auto flip_if = [&flags](bool cond, ::bson_validate_flags_t flag) {
```
Comment (acmorrow) [inline src/bsoncxx/validate.cpp]: can be const

File: src/bsoncxx/validate.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        validator._impl->_invalid_offset = stdx::make_optional(offset);
+        return {};
+    }
+
+    if (!::bson_validate(&bson, flags, &offset)) {
+        validator._impl->_invalid_offset = stdx::make_optional(offset);
+        return {};
+    }
+
+    return stdx::make_optional(document::view{data, length});
```
Comment (acmorrow) [inline src/bsoncxx/validate.cpp]: Can this be make_optional<Document>(data, length)?

**Comment by amidvidy** [comment]
@acmorrow PTAL?

**Comment by acmorrow** [comment]
LGTM. Next ABI rev maybe we will have a standards conformant Expected and we can get rid of the output parameter.

**Pull request closed without merging**